lua/impatient/cachepack.lua

```diff
@@ -0,0 +1,90 @@
+local ffi = require('ffi')
+
+local M = {}
+
+-- using double for packing/unpacking numbers has no conversion overhead
+local c_double = ffi.typeof("double[1]")
+local sizeof_c_double = ffi.sizeof("double")
+
+local OutputBuffer = {}
+
+function OutputBuffer.create() return {} end
+
+function OutputBuffer.write_number(buf, num)
+  buf[#buf+1] = ffi.string(c_double(num), sizeof_c_double)
+end
+
+function OutputBuffer.write_string(buf, str)
+  OutputBuffer.write_number(buf, #str)
+  buf[#buf+1] = str
+end
+
+function OutputBuffer.to_string(buf)
+  return table.concat(buf)
+end
+
+local InputBuffer = {}
+
+function InputBuffer.create(str)
+  return {
+    ptr = ffi.new("const char[?]", #str, str),
+    pos = 0,
+    size = #str
+  }
+end
+
+function InputBuffer.read_number(buf)
+  if (buf.size < buf.pos) then error("buffer access violation") end
+  local res = ffi.cast("double*", buf.ptr + buf.pos)[0]
+  buf.pos = buf.pos + sizeof_c_double
+  return res
+end
+
+function InputBuffer.read_string(buf)
+  local len = InputBuffer.read_number(buf)
+  local res = ffi.string(buf.ptr + buf.pos, len)
+  buf.pos = buf.pos + len
+
+  return res
+end
+
+function M.pack(cache)
+  _G.__luacache.used_mpack = false
+  local total_keys = vim.tbl_count(cache)
+  local buf = OutputBuffer.create()
+
+  OutputBuffer.write_number(buf, total_keys)
+  for k,v in pairs(cache) do
+    OutputBuffer.write_string(buf, k)
+    OutputBuffer.write_string(buf, v[1] or "")
+    OutputBuffer.write_number(buf, v[2] or 0)
+    OutputBuffer.write_string(buf, v[3] or "")
```

Comment:
Plenary messes with the cache directly, for example, so the cache can't be trusted to be in good form.

Serializing nil is pointless since it shouldn't be there in the first place, this is just to prevent a hard error that can't be recovered by pcall.